whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,213 @@
+        if (firstResource instanceof Ipv4Resource ipv4Resource){
+            final Ipv4Resource lastIpv4Resource = (Ipv4Resource)lastResource;
+            final Ipv4Resource parentIpv4Resource = (Ipv4Resource)parent;
+            return ipv4Resource.begin() == parentIpv4Resource.begin() && lastIpv4Resource.end() == parentIpv4Resource.end();
+        }
+
+        final Ipv6Resource ipv6Resource = (Ipv6Resource)firstResource;
+        final Ipv6Resource lastIpv6Resource = (Ipv6Resource)lastResource;
+        final Ipv6Resource parentIpv6Resource = (Ipv6Resource)parent;
+        return Objects.equals(ipv6Resource.begin(), parentIpv6Resource.begin()) && Objects.equals(lastIpv6Resource.end(), parentIpv6Resource.end());
+    }
+
+    private static List<IpEntry> findSiblingsAndExact(final IpTree ipTree, final IpInterval parentResource) {
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(parentResource);
+        if (parentList.isEmpty()){
+            return ipTree.findExact(parentResource);
+        }
+        return ipTree.findFirstMoreSpecific(IpInterval.parse(parentList.getFirst().getKey().toString()));
+    }
+
+    private IpEntry searchFirstLessSpecific(final IpTree ipTree, final IpInterval reverseIp){
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(reverseIp);
+        if (parentList.isEmpty() || !resourceExist(reverseIp, parentList.getFirst())){
+            throw new RdapException("404 Not Found", "No up level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+        return parentList.getFirst();
+    }
+
+    private List<IpEntry> searchTopLevelResource(final IpTree ipTree, final IpInterval reverseIp){
+        IpEntry ipEntry;
+        try {
+            ipEntry = searchFirstLessSpecific(ipTree, reverseIp);
+        } catch (RdapException ex){
+            throw new RdapException("404 Not Found", "No top-level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+
+        return List.of(loopUpLevels(ipTree, ipEntry));
+    }
+
+    private IpEntry loopUpLevels(final IpTree ipTree, IpEntry reverseIp) {
+        try {
+            reverseIp = searchFirstLessSpecific(ipTree, (IpInterval) reverseIp.getKey());
+            loopUpLevels(ipTree, reverseIp);
+        } catch (RdapException ex){
+            /*
+            * Do Nothing, end of loop
+            * */
+        }
+        return reverseIp;
+    }
+
+    private boolean resourceExist(final IpInterval ip, final IpEntry firstLessSpecific){
+        final RpslObject children = getResourceByKey(ip.toString());
+        final RpslObject rpslObject = getResourceByKey(firstLessSpecific.getKey().toString());
+        if (rpslObject == null || isAdministrativeResource(children, rpslObject)) {
+            LOGGER.debug("INET(6)NUM {} does not exist in RIPE Database ", firstLessSpecific.getKey().toString());
```

Comment:
can children be null to ?

Reply:
the children was the parent of the previous iteration.
So if anyone delete that entry from database in the same ms that someone is doing this call could happen yes.
I will add a validation just to cover all the possibilities